packages/controllers/src/permissions/PermissionController.ts

```diff
@@ -0,0 +1,1915 @@
+ * Consumers should use selector subscriptions to subscribe to relevant
+ * substate.
+ */
+export type PermissionControllerEvents = PermissionControllerStateChange;
+
+/**
+ * The external {@link ControllerMessenger} actions available to the
+ * {@link PermissionController}.
+ */
+type AllowedActions =
+  | AddApprovalRequest
+  | HasApprovalRequest
+  | AcceptApprovalRequest
+  | RejectApprovalRequest;
+
+/**
+ * The messenger of the {@link PermissionController}.
+ */
+export type PermissionControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  PermissionControllerActions | AllowedActions,
+  PermissionControllerEvents,
+  AllowedActions['type'],
+  never
+>;
+
+/**
+ * A generic {@link PermissionController}.
+ */
+export type GenericPermissionController = PermissionController<
+  PermissionSpecificationConstraint,
+  CaveatSpecificationConstraint
+>;
+
+/**
+ * Options for {@link PermissionController.grantPermissions}.
+ */
+type GrantPermissionsOptions = {
+  approvedPermissions: RequestedPermissions;
+  subject: PermissionSubjectMetadata;
+  preserveExistingPermissions?: boolean;
+  requestData?: Record<string, unknown>;
+};
+
+/**
+ * Describes the possible results of a {@link CaveatMutator} function.
+ */
+export enum CaveatMutatorOperation {
+  noop,
+  updateValue,
+  deleteCaveat,
+  revokePermission,
+}
+
+/**
+ * Given a caveat value, returns a {@link CaveatMutatorOperation} and, optionally,
+ * a new caveat value.
+ *
+ * @see {@link PermissionController.updatePermissionsByCaveat} for more details.
+ * @template Caveat - The caveat type for which this mutator is intended.
+ * @param caveatValue - The existing value of the caveat being mutated.
+ * @returns A tuple of the mutation result and, optionally, the new caveat
+ * value.
+ */
+export type CaveatMutator<TargetCaveat extends CaveatConstraint> = (
+  caveatValue: TargetCaveat['value'],
+) =>
+  | [Exclude<CaveatMutatorOperation, CaveatMutatorOperation.updateValue>]
+  | [CaveatMutatorOperation.updateValue, CaveatConstraint['value']];
+
+/**
+ * Extracts the permission(s) specified by the given permission and caveat
+ * specifications.
+ *
+ * @template ControllerPermissionSpecification - The permission specification(s)
+ * to extract from.
+ * @template ControllerCaveatSpecification - The caveat specification(s) to
+ * extract from. Necessary because {@link Permission} has a generic parameter
+ * that describes the allowed caveats for the permission.
+ */
+export type ExtractPermission<
+  ControllerPermissionSpecification extends PermissionSpecificationConstraint,
+  ControllerCaveatSpecification extends CaveatSpecificationConstraint,
+> = ControllerPermissionSpecification extends ValidatePermissionSpecification<ControllerPermissionSpecification>
+  ? Permission<
+      ControllerPermissionSpecification['targetKey'],
+      ExtractCaveats<ControllerCaveatSpecification>
+    >
+  : never;
```

Comment:
Nit: the `ValidationPermissionSpecification` type can be used directly on the generic param, to simplify this a little bit

```suggestion
export type ExtractPermission<
  ControllerPermissionSpecification extends ValidatePermissionSpecification<PermissionSpecificationConstraint>,
  ControllerCaveatSpecification extends CaveatSpecificationConstraint,
> = Permission<
  ControllerPermissionSpecification['targetKey'],
  ExtractCaveats<ControllerCaveatSpecification>
>;
```